lib/utils/normalization.ts

```diff
@@ -27,25 +28,58 @@ export function normalizeUuidProp(
 	}
 }
 
+/**
+ * Takes a string which may represent one of:
+ * 	- Integer release id
+ *  - String commit hash, full 32char
+ *  - String commit hash, short
+ * and returns the correctly typed value (integer|string).
+ * @param balena balena sdk
+ * @param release string representation of release reference (id/hash)
+ */
 export async function disambiguateReleaseParam(
 	balena: BalenaSDK,
-	param: string | number,
-	paramRaw: string | undefined,
+	release: string,
 ) {
-	// the user has passed an argument that was parsed as a string
-	if (typeof param !== 'number') {
-		return param;
+	// Reject empty release or invalid characters
+	const nonEmptyAlphaNumeric = /^[a-z0-9]+$/;
+	if (!release || release.length === 0 || !nonEmptyAlphaNumeric.test(release)) {
+		throw new ExpectedError('Invalid release parameter');
 	}
 
-	// check whether the argument was indeed an ID
-	return balena.models.release
-		.get(param, { $select: 'id' })
-		.catch(error => {
-			// we couldn't find a release by id,
-			// try whether it was a commit with all numeric digits
-			return balena.models.release
-				.get(paramRaw || _.toString(param), { $select: 'id' })
-				.catchThrow(error);
-		})
-		.then(({ id }) => id);
+	const possibleHashLength = release.length === 7 || release.length === 32;
+	const hasLeadingZero = release[0] === '0';
+	const isOnlyNumerical = /^[0-9]+$/.test(release);
+
+	// Reject non-numerical values with invalid hash lengths
+	if (!isOnlyNumerical && !possibleHashLength) {
+		throw new ExpectedError('Invalid release parameter');
+	}
+
+	// Reject leading-zero values with invalid hash lengths
+	if (hasLeadingZero && !possibleHashLength) {
+		throw new ExpectedError('Invalid release parameter');
+	}
+
+	// If alphanumeric, must be hash.
+	if (!isOnlyNumerical) {
+		return release;
+	}
+
+	// Now very likely an integer id (but still could be number only hash)
+	// Check integer id with API
+	try {
+		return (
+			await balena.models.release.get(parseInt(release, 10), {
+				$select: 'id',
+			})
+		).id;
+	} catch (e) {
+		if (e.name !== 'BalenaReleaseNotFound') {
+			throw e;
+		}
+	}
+
+	// Must be a number only hash (or nonexistent release)
```

Comment:
Not sure I understand the first point - it seems to me that `tryAsInteger` handles e.g. '012345' correctly - i.e. returns it as a string.

In order to get rid of tryAsInteger() we'd need to add `disambiguateDevice`, `disambiguateApplication` etc, since they use the sdk methods specific to the context.  I do think that we should consider `tryAsInteger` a bit of a hack though, but I think we should consider the whole DB id thing before getting rid of it.

Reply:
I mean that if we apply `tryAsInteger` to a device UUID such as in:
```
balena.models.device.tags.set(tryAsInteger(options.device), ...)
```
then it would make sense to apply the hash length tests of `disambiguateReleaseParam` there too, because a device UUID can only be 7, 32 or 62 characters long. So my suggestion is to try and come up with a single parsing function for hashes (UUID or commit hash), instead of having multiple functions like `tryAsInteger`, `disambiguateDevice`, `disambiguateReleaseParam`, etc. OK, `application` may be different because it is either a name like "test-rpi" or a database ID like "1234", so perhaps the existing `tryAsInteger` test is OK (#1842 applies/remains). Maybe the reusable function would be "isHash", so the hash length tests are shared. Not sure how much the database querying logic could be shared, just replacing the resource name (`release`, `device`...).
```
export async function isHash(
	hash: string,
	allowCommitHash = false,
) {
```

... and let only the database querying bit be specific to release, device, etc.

> Not sure I understand the first point

Specifically that `disambiguateReleaseParam`, as implemented, allows a numeric database ID to be specified with a leading zero, say `0123456` (it will end up calling `parseInt('0123456')` and querying the API for database ID `123456`), whereas `tryAsInteger` assumes that a numeric database ID cannot be specified with a leading zero. It would be better to be consistent on this "rule".